# Here in this code, i have a dataset that contain data about many features of tumors, and i will be making a machine learning pipeline for predicting if cancer is malginent or benign, to help doctors in their tests

First i will upload the libraries we will need to upload the dataset and split the data and test and train the data

In [ ]:
import pandas as pd
import sklearn.metrics
import sklearn.model_selection

then i will upload the dataset

knowing that there are many columns, i will use 'display.max_columns' option to display all of them and look for any non numeric or categorial data, if any non numeric exist i must encode it

In [ ]:
pd.set_option('display.max_columns', None)
df = pd.read_csv("Cancer_Data.csv")
df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     symmetry_mean  fractal_dimension_mean  radius_se  texture_se  \
0           0.2419                 0.07871     1.0950      0.9053   
1           0.1812                 0.05667     0.5435      0.7339   
2           0.2069                 0.05999     0.7456      0.7869   
3           0.2597                 0.09744     0.4956      1.1560   
4           0.1809                 0.05883     0.7572      0.7813   
..             ...                     ...        ...         ...   
564         0.1726                 0.05623     1.1760      1.2560   
565         0.1752                 0.05533     0.7655      2.4630   
566         0.1590                 0.05648     0.4564      1.0750   
567         0.2397                 0.07016     0.7260      1.5950   
568         0.1587                 0.05884     0.3857      1.4280   

     perimeter_se  area_se  smoothness_se  compactness_se  concavity_se  \
0           8.589   153.40       0.006399         0.04904       0.05373   
1           3.398    74.08       0.005225         0.01308       0.01860   
2           4.585    94.03       0.006150         0.04006       0.03832   
3           3.445    27.23       0.009110         0.07458       0.05661   
4           5.438    94.44       0.011490         0.02461       0.05688   
..            ...      ...            ...             ...           ...   
564         7.673   158.70       0.010300         0.02891       0.05198   
565         5.203    99.04       0.005769         0.02423       0.03950   
566         3.425    48.55       0.005903         0.03731       0.04730   
567         5.772    86.22       0.006522         0.06158       0.07117   
568         2.548    19.15       0.007189         0.00466       0.00000   

     concave points_se  symmetry_se  fractal_dimension_se  radius_worst  \
0              0.01587      0.03003              0.006193        25.380   
1              0.01340      0.01389              0.003532        24.990   
2              0.02058      0.02250              0.004571        23.570   
3              0.01867      0.05963              0.009208  

the values i saw for the last column (Unnamed: 32) are all NaN, so i will check if all of them are like this, if their is similar to the number of rows then i will delete the column

In [ ]:
dfnull = df['Unnamed: 32'].isna().sum()
dfnull

569

569 is equal to the number of rows, so i will delete the column and also the id column because it's not needed in the training and don't offer any information.

In [ ]:
df = df.drop(['id', 'Unnamed: 32'], axis=1)

Here i want to check if all the values in the diagnosis column are malignent or benign (no corrupted data in the column), i will count how many times both values are mentioned and then see it's equal to the number of total rows

In [ ]:
C1 = (df["diagnosis"] == 'M').sum()
C2 = (df["diagnosis"] == 'B').sum()
print(C1)
print(C2)

212
357


then i will turn them to 1 and 0, so that they could be used for the trainig and testing of models

In [ ]:
df['diagnosis'] = df['diagnosis'] .replace({'B': 1, 'M': 0})

Here i will split the dataset to train and test sets, and since the data isn't very large, i will take less data for test than 20%

In [ ]:
df_train, df_test = sklearn.model_selection.train_test_split(df,test_size=0.15)
print("df:", df.shape)
print("df_train:", df_train.shape)
print("df_test:", df_test.shape)

df: (569, 31)
df_train: (483, 31)
df_test: (86, 31)


# Preprocess

First i will check for empty values to remove them in the training set

In [ ]:
dfnull = df_train.isna().sum()
dfnull

diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64

then i will check for duplicate rows, and i will delete if any exists

In [ ]:
dfduplicate =df_train.duplicated().sum()
dfduplicate

0

Here i will check for outliers to delete, the dataset isn't very big so i will try to delete less number of the data

but here i will only pick the columns with big differece between their min and max values.

first i will check the max and min values in them, then remove the outliers

In [ ]:
minval = df_train.min()
maxval = df_train.max()
print(minval)
print(maxval)

diagnosis                    0.000000
radius_mean                  6.981000
texture_mean                 9.710000
perimeter_mean              43.790000
area_mean                  143.500000
smoothness_mean              0.052630
compactness_mean             0.019380
concavity_mean               0.000000
concave points_mean          0.000000
symmetry_mean                0.106000
fractal_dimension_mean       0.049960
radius_se                    0.111500
texture_se                   0.360200
perimeter_se                 0.757000
area_se                      6.802000
smoothness_se                0.001713
compactness_se               0.002252
concavity_se                 0.000000
concave points_se            0.000000
symmetry_se                  0.007882
fractal_dimension_se         0.000950
radius_worst                 7.930000
texture_worst               12.020000
perimeter_worst             50.410000
area_worst                 185.200000
smoothness_worst             0.071170
compactness_

I saw that the columns that have biggest differences between their min amd max values are [area_mean,area_se,area_worst], so i will only remove the outlier values in them

In [ ]:
Q1 = df_train['area_mean'].quantile(0.10)
Q3 = df_train['area_mean'].quantile(0.90)
IQR = Q3 - Q1

df_train = df_train[(df_train['area_mean'] >= Q1 - 1.5 * IQR) & (df_train['area_mean'] <= Q3 + 1.5 * IQR)]

In [ ]:
Q1 = df_train['area_se'].quantile(0.10)
Q3 = df_train['area_se'].quantile(0.90)
IQR = Q3 - Q1

df_train = df_train[(df_train['area_se'] >= Q1 - 1.5 * IQR) & (df_train['area_se'] <= Q3 + 1.5 * IQR)]

In [ ]:
Q1 = df_train['area_worst'].quantile(0.10)
Q3 = df_train['area_worst'].quantile(0.90)
IQR = Q3 - Q1

df_train = df_train[(df_train['area_worst'] >= Q1 - 1.5 * IQR) & (df_train['area_worst'] <= Q3 + 1.5 * IQR)]

In [ ]:
df_train.shape

(479, 31)

and here i will split the train and test data for a sets for the the target label, and a sets for the other feautres

In [ ]:
x_train = df_train.drop(["diagnosis"], axis = 1)
y_train = df_train["diagnosis"]

x_test = df_test.drop(["diagnosis"], axis = 1)
y_test = df_test["diagnosis"]

print("x train size:", x_train.shape)
print("x test size:", x_test.shape)
print("y train size:", y_train.shape)
print("y test size:", y_test.shape)

x train size: (479, 30)
x test size: (86, 30)
y train size: (479,)
y test size: (86,)


# Train, Evaluate and test

Here i will try to find the best model to train it and then test it
i must upload the libriaries i need for the training and testing of the model
I will use several ensemble learning methods and then pick the best using gridsearch

first i will upload the libraries needed for the
and ensemble learning methods

In [ ]:
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score, confusion_matrix
from sklearn.model_selection import GridSearchCV

the ensemble learning algorithms i picked are random forest for it's accuracy and handling complex relations between features, Extra trees because it's faster and the dataset have many dimensions, ADA because it's effective for binary classification, Voting classifier to use severla methods at once

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import VotingClassifier

I will tune the hyperparameters for the Randomforestclassifer, i will set criterion for different ways to test impurity of splits, i will pick the default and a lower value for the n_estimator (since our dataset isn't very large)

In [ ]:
RFC_parameters_grid = {"criterion" : ["gini", "entropy", "log_loss"],
                   "n_estimators": ( 50, 100),
                        "max_depth": (5,10,None)}
RFC_model = sklearn.model_selection.GridSearchCV(sklearn.ensemble.RandomForestClassifier(),
             RFC_parameters_grid, scoring="accuracy")
RFC_model.fit(x_train, y_train)
print ("accuracy {:.2f}".format(RFC_model.best_score_))
print ("best found hyperparameters  = {}".format(RFC_model.best_params_))

accuracy 0.96
best found hyperparameters  = {'criterion': 'log_loss', 'max_depth': 10, 'n_estimators': 100}


for ada boosting classifier i will also pick some lower values with the default ones for learning rate and n_estimators, because the dataset isn't large, also different algoritms

In [ ]:
ada_parameters_grid = { "learning_rate" : [ 0.1, 0.2, 1.0],
                   "n_estimators": (20, 50),
                   "algorithm": ('SAMME', 'SAMME.R')}
ADAmodel = sklearn.model_selection.GridSearchCV(sklearn.ensemble.AdaBoostClassifier(),
             ada_parameters_grid, scoring="accuracy")
ADAmodel.fit(x_train, y_train)
print ("accuracy {:.2f}".format(ADAmodel.best_score_))
print ("best found hyperparameters = {}".format(ADAmodel.best_params_))

accuracy 0.96
best found hyperparameters = {'algorithm': 'SAMME', 'learning_rate': 1.0, 'n_estimators': 50}


in Extra tree classifier, i also use default/lower values for n_estimators and max depth, several criterion methods,and raise minumum decrease of impurity needed to split a node and allow out of bag samples

In [ ]:
ETC_parameters_grid = { "n_estimators": [50,100],
                       "criterion": ['gini', 'entropy'],
                        "max_depth": [None,5,10],
                       "min_impurity_decrease":[0.0, 0.5],
                   "oob_score": (True, False)}
ETCmodel = sklearn.model_selection.GridSearchCV(sklearn.ensemble.ExtraTreesClassifier(),
             ETC_parameters_grid, scoring="accuracy")
ETCmodel.fit(x_train, y_train)
print ("accuracy {:.2f}".format(ETCmodel.best_score_))
print ("best found hyperparameters = {}".format(ETCmodel.best_params_))

accuracy 0.97
best found hyperparameters = {'criterion': 'gini', 'max_depth': None, 'min_impurity_decrease': 0.0, 'n_estimators': 50, 'oob_score': False}


/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
120 fits failed out of a total of 240.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
120 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py", line 434, in fit
    raise ValueError("Out of bag estimation only available if bootstrap=True")
ValueError: Out of bag estimation only available if bootstrap=True

  warnings.warn(some_fits_failed_message, FitFailedWarning)
/usr/local/

for the voting classifier, i must first upload and use some classificaiton algorithms and tune them

For svm/svc, i will try 3 different kernels

for logistic regression i will try using all processors and or just one

for Multinomial naiv Bayes, i will use prior probabilities or not

for the decision tree, to measure quality of the tree spliting i will try gini and entropy
different tree depths, and numbers required for split to avoid overfitting
put some values for min_impurity_decrease to put more restriction for splits in the tree by putting minimum amount of impurity decrease to split a node

for the Voting Classifier i will use both hard and soft voting

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression

T = DecisionTreeClassifier()
S = SVC()
MNB = MultinomialNB()
L = LogisticRegression()

T_param_grid = {'criterion': ['gini', 'entropy'], 'max_depth': [3, 5, 10],
                'min_samples_split': [1,2],'min_impurity_decrease': [1,2,0.0]}
S_param_grid = {'kernel': ['linear', 'rbf', 'poly']}
L_param_grid = { 'n_jobs': [None, -1]}
MNB_param_grid = {'fit_prior': [True,False]}

In [ ]:
estimations = [('dt', T), ('svm',S), ('NB',MNB), ('LR',L)]
VC_parameters_grid = { "voting": ['hard', 'Soft']}
VCmodel = sklearn.model_selection.GridSearchCV(sklearn.ensemble.VotingClassifier(estimations),
             VC_parameters_grid, scoring="accuracy")
VCmodel.fit(x_train, y_train)
print ("accuracy {:.2f}".format(VCmodel.best_score_))
print ("best found hyperparameters = {}".format(VCmodel.best_params_))

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

accuracy 0.95
best found hyperparameters = {'voting': 'hard'}


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
5 fits failed out of a total of 10.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
5 fits failed with the following error:
Traceback

and here i want to pick the best model

In [ ]:
Bestmodel =  RFC_model.best_estimator_ if RFC_model.best_score_ > ADAmodel.best_score_ and RFC_model.best_score_ > ETCmodel.best_score_ and RFC_model.best_score_ > VCmodel.best_score_ else ADAmodel.best_estimator_ if ADAmodel.best_score_ > ETCmodel.best_score_ and ADAmodel.best_score_ > VCmodel.best_score_ else ETCmodel.best_estimator_ if ETCmodel.best_score_ > VCmodel.best_score_ else VCmodel.best_estimator_
Bestmodel.fit(x_train, y_train)

ExtraTreesClassifier(n_estimators=50)

and here i will see the result of the model on the test set

In [ ]:
y_pred = Bestmodel.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
CM = confusion_matrix(y_test, y_pred)

print("accuracy:", accuracy)
print("confusion matrix:\n", CM)
print("FI:", f1)
print("Precision:", precision)
print("recall:", recall)

accuracy: 0.9534883720930233
confusion matrix:
 [[28  4]
 [ 0 54]]
FI: 0.9642857142857143
Precision: 0.9310344827586207
recall: 1.0


#Discussion:


We got some good results at the end of this pipeline, with few notes:


The accuracy number is very well, when we see the recall number, it suggests that the model is doing perfect with positive instnaces (predicting Malignent tumor) and labeled as such in the dataset, precision shows that almost all positives were corectly predicted, F1 score show us that the model have great performance becasue it's close to 1.


as for the confusion matrix, the model also seems to be almost perfect, and it was able to predict all positive instances correctly


but we have to take into account that there is a gap (but not big) between the count of positive and negative instances, with Benign being more, we can conclude that there maybe still need for some experimentation by adding or removing data by appling data augmentation for malignent instances,or dimensionality reduction on Benign instnaces.


for later improvement, we can consider adding or removing more data, and also having different sizes for the test set, however considering the dataset was small, traditional predicting models was possibly sufficent

# other experiments


_I also tried to use Perceptron (the code is at the end of this page) and we have a good result of around 0.85, but it seems testing many models was with ensemble and gridsearchCV gave use more options

_ i also replaced the quantile values for Q1 and Q3 of 0.10 and 0.90 with o.15 and 0.85, and i got an accuracy result of 0.94 with other metrics being also high but little bit less, which most likely mean it was better to keep as much data as possible

Dataset refrencing:

Taha, E. (2023) Cancer data, Kaggle. Available at: https://www.kaggle.com/datasets/erdemtaha/cancer-data (Accessed: 27 June 2023).